modules/services/yabai/default.nix

```diff
@@ -0,0 +1,107 @@
+{ config, lib, pkgs, ... }:
+
+with lib;
+
+let
+  cfg = config.services.yabai;
+
+  toYabaiConfig = opts:
+    concatStringsSep "\n" (mapAttrsToList
+      (p: v: "yabai -m config ${p} ${toString v}") opts);
+
+in
+
+{
+  options = {
+    services.yabai = with types; {
+      enable = mkOption {
+        type = bool;
+        default = false;
+        description = "Whether to enable the yabai window manager.";
+      };
+
+      package = mkOption {
+        type = path;
+        default = pkgs.yabai;
+        description = "The yabai package to use.";
+      };
+
+      enableScriptingAddition = mkOption {
+        type = bool;
+        default = false;
+        description = ''
+          Whether to enable yabai's scripting-addition.
+          SIP must be disabled for this to work.
+        '';
+      };
+
+      config = mkOption {
+        type = attrs;
+        default = {};
+        example = literalExample ''
+          {
+            focus_follows_mouse = "autoraise";
+            mouse_follows_focus = "off";
+            window_placement    = "second_child";
+            window_opacity      = "off";
+            top_padding         = 36;
+            bottom_padding      = 10;
+            left_padding        = 10;
+            right_padding       = 10;
+            window_gap          = 10;
+          }
+        '';
+        description = ''
+          Key/Value pairs to pass to yabai's 'config' domain, via the configuration file.
+        '';
+      };
+
+      extraConfig = mkOption {
+        type = str;
+        default = "";
+        example = literalExample ''
+          yabai -m rule --add app='System Preferences' manage=off
+        '';
+        description = "Extra arbitrary configuration to append to the configuration file";
+      };
+    };
+  };
+
+  config = mkMerge [
+    (mkIf (cfg.enable) {
+      environment.systemPackages = [ cfg.package ];
+
+      launchd.user.agents.yabai = {
+        serviceConfig.ProgramArguments = [ "${cfg.package}/bin/yabai" ]
+                                         ++ optionals (cfg.config != {} || cfg.extraConfig != "") [ "-c" "/etc/yabairc"];
+        serviceConfig.KeepAlive = true;
+        serviceConfig.RunAtLoad = true;
+        serviceConfig.EnvironmentVariables = {
+          PATH = "${cfg.package}/bin:/usr/bin:/bin:/usr/sbin:/sbin";
+        };
+      };
+    })
+
+    (mkIf (cfg.config != {} || cfg.extraConfig != "") {
+      environment.etc."yabairc".source = pkgs.writeScript "etc-yabairc" (
```

Comment:
Any point in exposing this in `/etc`?

Reply:
Common convention that’s already adhered to in several other modules in this codebase

Reply:
If the service supports hot reloading for example /etc can be used to update the configuration in a stable location without restarting, otherwise it's one more thing that doesn't have to be global.

Reply:
Okay :) So perhaps rather than writing `/etc/yabairc` it'd just be an arbitrary path? Something like this?
```nix
{
      launchd.user.agents.yabai = {
        serviceConfig.ProgramArguments = [ "${cfg.package}/bin/yabai" ]
                                         ++ optionals (cfg.config != {} || cfg.extraConfig != "")
                                           [ "-c" pkgs.writeScript "yabairc" (
                                             (if (cfg.config != {})
                                              then "${toYabaiConfig cfg.config}"
                                              else "")
                                             + optionalString (cfg.extraConfig != "") cfg.extraConfig)];
        serviceConfig.KeepAlive = true;
        serviceConfig.RunAtLoad = true;
        serviceConfig.EnvironmentVariables = {
          PATH = "${cfg.package}/bin:/usr/bin:/bin:/usr/sbin:/sbin";
        };
      };
}
```

Reply:
Yeah, alltho would be nicer to put this in the let binding at the top and use `[ "-c" configFile ]` so it's a bit more readable

Reply:
Good idea 👍 I've updated it like so  
Just wondering how we do the tests now? As I'd written them to grep for strings in the resulting `/etc/yabairc`, but since that path is now unknown, I'm not sure how to approach this

Reply:
Not sure if this is really the right approach here, but I got the test passing by using `sed` to pull out the store path from the plist:
```
  test = ''
    echo >&2 "checking yabai service in ~/Library/LaunchAgents"
    grep "org.nixos.yabai" ${config.out}/user/Library/LaunchAgents/org.nixos.yabai.plist
    grep "${yabai}/bin/yabai" ${config.out}/user/Library/LaunchAgents/org.nixos.yabai.plist

    conf=`sed -En '/<string>-c<\/string>/{n; s/\s+?<\/?string>//g; p;}' ${config.out}/user/Library/LaunchAgents/org.nixos.yabai.plist`
    echo >&2 "checking config in $conf"
    grep "yabai -m config focus_follows_mouse autoraise" $conf
    grep "yabai -m rule --add app='System Preferences' manage=off" $conf
  '';
```
Tested with `nix-build release.nix -A tests.services-yabai` and it works.  
Thoughts @LnL7? Perhaps you know of a better way to do this?

Reply:
I squashed in that change [here](https://github.com/LnL7/nix-darwin/commit/7ef533e8d1fb6ece563232b326fc5b5d5c2f9257#diff-e566c25b87e15708d8368b82febe145cR20) so you can see (can remove if there's a nicer way, of course)

Reply:
Quite a few nixos modules also have a [configFile](https://github.com/NixOS/nixpkgs/blob/release-20.03/nixos/modules/config/pulseaudio.nix#L119) option, which you could reference in the tests to access the file path. But this is also fine, the tests are more a sanity check anyway.

Reply:
If you’re fine with this, perhaps let’s just keep it this way. If there’s interest in the future for a configFile option, I can add it :)

Reply:
Sounds good